internal/services/admin/server.go

```diff
@@ -646,3 +705,224 @@ func toWorkflowVersion(workflowVersion *dbsqlc.GetWorkflowVersionForEngineRow) *
 
 	return version
 }
+
+func getOpts(ctx context.Context, requests []*contracts.TriggerWorkflowRequest, a *AdminServiceImpl) ([]*repository.CreateWorkflowRunOpts, []string, error) {
+	tenant := ctx.Value("tenant").(*dbsqlc.Tenant)
+	tenantId := sqlchelpers.UUIDToStr(tenant.ID)
+
+	results := make([]*repository.CreateWorkflowRunOpts, 0)
+	existingWorkflowRuns := make([]string, 0)
+
+	nonParentWorkflows := make([]*contracts.TriggerWorkflowRequest, 0)
+
+	createContext, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+	defer cancel()
+
+	for _, req := range requests {
+		isParentTriggered := req.ParentId != nil
+
+		if isParentTriggered {
+			if req.ParentStepRunId == nil {
+				return nil, nil, status.Error(
+					codes.InvalidArgument,
+					"parent step run id is required when parent id is provided",
+				)
+			}
+
+			if req.ChildIndex == nil {
+				return nil, nil, status.Error(
+					codes.InvalidArgument,
+					"child index is required when parent id is provided",
+				)
+			}
+
+			workflowRun, err := a.repo.WorkflowRun().GetChildWorkflowRun(
```

Comment:
This is an `n+1` style query. This should probably be a bulk list operation as well with `(parentId, parentStepRunId, index, key)` tuples to match the where condition.

Reply:
ok, wasn't clear to me how often we were hitting these paths

Reply:
also this is quite a bit more complicated - if we don't find the child workflow we have to trigger the workflow, I could key a hash by the params we use - but it's starting to get even more messy so I'm not sure if the juice is worth the squeeze.

Reply:
ok added a fix for this